### PR TITLE
Reorganization of histos in Phase2TK DQM digi monitoring

### PR DESCRIPTION
**Pull request by gsaha009**
#### PR description:

Major Changes:
  - In the endcap, disc level histogramming has been changed to ring level 
  - Overflow and range of some histograms are fixed

#### PR validation:
 runTheMatrix test workflows : 23234.0, 23234.21 (tested in latest CMSSW release : CMSSW_11_2_X_2020-09-22-2300)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

NA

### CODE REVIEW

**Comment by mmusich** [comment]
@gsaha009, thanks for the PR.
Can you edit the title and use a more self explanatory one? Thanks

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 1]
```diff
@@ -445,32 +433,30 @@ void Phase2TrackerMonitorDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
-    std::ostringstream fname1, fname2, tag;
+    std::ostringstream fname1;
     int side = 0;
     int idisc = 0;
-    if (layer < 100) {
-      fname1 << "Barrel";
-      fname2 << "Layer_" << layer;
-    } else {
+
+    //    std::string key = getHistoId(det_id, tTopo, pixelFlag_);
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:448]: Remove commented out code

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 2]
```diff
@@ -699,25 +691,76 @@ void Phase2TrackerMonitorDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
+
+  if (layer < 100) {
+    fname1 << "Barrel/";
+    fname1 << "Layer" << layer;
+    fname1 << "";
+  } else {
+    int side = layer / 100;
+    fname1 << "EndCap_Side" << side << "/";
+    int disc = layer - side * 100;
+    if (flag) Disc = (disc < 9) ? "FPIX_1" : "FPIX_2";
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:738]: Perhaps EPix (endcap pixel) and FPix (forward pixel) are more inline with official nomemclature.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 3]
```diff
@@ -699,25 +691,76 @@ void Phase2TrackerMonitorDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
+    layerMEs.insert(std::make_pair(key, local_mes));
+  }
+}
+std::string Phase2TrackerMonitorDigi::getHistoId(uint32_t det_id, const TrackerTopology* tTopo, bool flag) {
+  int layer;
+  std::string Disc;
+  std::ostringstream fname1;
+  if(flag){
+    layer = tTopo->getITPixelLayerNumber(det_id);
+    //std::cout<<"PxLayer: "<<layer<<"\n";
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:722]: Remove commented out code.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 4]
```diff
@@ -47,6 +47,7 @@
 Phase2TrackerMonitorDigi::Phase2TrackerMonitorDigi(const edm::ParameterSet& iConfig)
     : config_(iConfig),
       pixelFlag_(config_.getParameter<bool>("PixelPlotFillingFlag")),
+      clsFlag_(config_.getParameter<bool>("OfflineClusteriserFlag")),
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:50]: the usage of "Offline" in clusterizer may be misleading here. If I understand correctly this are clusters remade by scratch from the digis (so some sort of home-brew clusters), no?

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 5]
```diff
@@ -699,25 +691,76 @@ void Phase2TrackerMonitorDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
+    fname1 << "Layer" << layer;
+    fname1 << "";
+  } else {
+    int side = layer / 100;
+    fname1 << "EndCap_Side" << side << "/";
+    int disc = layer - side * 100;
+    if (flag) Disc = (disc < 9) ? "FPIX_1" : "FPIX_2";
+    else Disc = (disc < 3) ? "TEDD_1" : "TEDD_2";
+    fname1 << Disc << "/";
+    int ring = tTopo->tidRing(det_id);
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:741]: I am somewhat puzzled by the logic here. How is the ring assinged in the case of the IT? ` tTopo->tidRing(det_id)` should work only for OT.

File: DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py [hunk 6]
```diff
@@ -2,136 +2,137 @@
...
+                        DigiChargeH = cms.PSet(
+                            Nbins = cms.int32(50),
+                            xmin   = cms.double(-0.5),
+                            xmax   = cms.double(49.5),
+                            switch = cms.bool(True)
+                        ), 
+                        TotalNumberOfDigisPerLayerH = cms.PSet(
+                            Nbins = cms.int32(1000),
+                            xmin   = cms.double(-0.5),
+                            xmax   = cms.double(4999.5),
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:72]: think in the PU case we can have much higher occupancies, see e.g: https://tinyurl.com/y4thv7xg
image

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 7]
```diff
@@ -166,7 +166,7 @@ void Phase2TrackerValidateDigi::analyze(const edm::Event& iEvent, const edm::Eve
     }
 
     fillSimHitInfo(iEvent, (*simTrkItr), geomHandle);
-    fillHitsPerTrack();
+    //fillHitsPerTrack();
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:169]: remove commented out code

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 8]
```diff
@@ -595,9 +595,9 @@ void Phase2TrackerValidateDigi::bookHistograms(DQMStore::IBooker& ibooker,
   else
     SimulatedTOFZMap = nullptr;
 
-  HistoName.str("");
-  HistoName << "NumberOfSimHitsPerTrackVsLayer";
-  nSimHitsPerTrack = ibooker.book2D(HistoName.str(), HistoName.str(), 45, -22.5, 22.5, 101, -0.5, 100.5);
+  //HistoName.str("");
+  //HistoName << "NumberOfSimHitsPerTrackVsLayer";
+  //nSimHitsPerTrack = ibooker.book2D(HistoName.str(), HistoName.str(), 45, -22.5, 22.5, 101, -0.5, 100.5);
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:600]: remove commented out code.

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 9]
```diff
@@ -657,45 +657,49 @@ void Phase2TrackerValidateDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
-    if (layer < 100) {
-      fname1 << "Barrel";
-      fname2 << "Layer_" << layer;
-    } else {
-      int side = layer / 100;
-      int idisc = layer - side * 100;
-      fname1 << "EndCap_Side_" << side;
-      fname2 << "Disc_" << idisc;
+    std::ostringstream fname1;
+    //    std::string key = getHistoId(det_id, tTopo, flag);
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:677]: remove commented out code

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 10]
```diff
@@ -657,45 +657,49 @@ void Phase2TrackerValidateDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
...
     ibooker.cd();
-    folder_name << top_folder << "/"
-                << "DigiMonitor"
-                << "/" << fname1.str() << "/" << fname2.str();
-    edm::LogInfo("Phase2TrackerValidateDigi") << " Booking Histograms in : " << folder_name.str();
+    ibooker.setCurrentFolder(top_folder+"/DigiMonitor/"+key);
+    edm::LogInfo("Phase2TrackerValidateDigi") << " Booking Histograms in : " << key;
 
-    ibooker.setCurrentFolder(folder_name.str());
+    //    ibooker.setCurrentFolder(folder_name.str());
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:695]: remove commented out code.

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 11]
```diff
@@ -1072,5 +1085,35 @@ void Phase2TrackerValidateDigi::fillHitsPerTrack() {
...
+
+  if (layer < 100) {
+    fname1 << "Barrel/";
+    fname1 << "Layer" << layer;
+    fname1 << "";
+  } else {
+    int side = layer / 100;
+    fname1 << "EndCap_Side" << side << "/";
+    int disc = layer - side * 100;
+    if (flag) Disc = (disc < 9) ? "FPIX_1" : "FPIX_2";
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:1110]: as above, perhaps EPix and FPix are more inline with the official nomenclature.

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc [hunk 12]
```diff
@@ -1072,5 +1085,35 @@ void Phase2TrackerValidateDigi::fillHitsPerTrack() {
...
+    else Disc = (disc < 3) ? "TEDD_1" : "TEDD_2";
+    fname1 << Disc << "/";
+    int ring = tTopo->tidRing(det_id);
+    fname1 << "Ring" << ring;
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:1114]: as above I am puzzled by how this works in the case of FPix.
Also this function is essentially a copy of the other one in `Phase2TrackerMonitorDigi.cc`. It would be cleaner to have some static method declared elsewhere and used in both classes.

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.h [hunk 13]
```diff
@@ -112,7 +113,8 @@ class Phase2TrackerValidateDigi : public DQMEDAnalyzer {
   void fillHitsPerTrack();
 
   edm::ParameterSet config_;
-  std::map<unsigned int, DigiMEs> layerMEs;
+  //  std::map<unsigned int, DigiMEs> layerMEs;
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.h:116]: remove commented out code.

File: Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py [hunk 14]
```diff
@@ -24,7 +24,8 @@
     SimTrackSource = cms.InputTag("g4SimHits"),
     SimVertexSource = cms.InputTag("g4SimHits"),
     GeometryType = cms.string('idealForDigi'),
-    PtCutOff      = cms.double(9.5),                           
+    #PtCutOff      = cms.double(9.5),
+    PtCutOff      = cms.double(2.0),
```
Comment (mmusich) [inline Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py:28]: what is the rationale behind this change of the pT cut?

**Review by mmusich — CHANGES_REQUESTED** [review]
@gsaha009,
Please apply the mandatory code-format and code-check patch as suggested from the bot: https://github.com/cms-sw/cmssw/pull/31563#issuecomment-698208738 otherwise the integration tests cannot be started.

Also I have some comments on the code, see below:

**Comment by mmusich** [comment]
@emiglior and @skinnari FYI.

**Comment by mmusich** [comment]
@cmsbuild, please test

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 15]
```diff
@@ -97,19 +98,25 @@ void Phase2TrackerMonitorDigi::fillITPixelDigiHistos(const edm::Handle<edm::DetS
   const TrackerTopology* tTopo = tTopoHandle_.product();
   const TrackerGeometry* tGeom = gHandle.product();
 
+  //  std::cout << " Here " << std::endl;
```
[the remark this answers is not in the document]
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:101]: Done

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 16]
```diff
@@ -97,19 +98,25 @@ void Phase2TrackerMonitorDigi::fillITPixelDigiHistos(const edm::Handle<edm::DetS
...
   for (typename edm::DetSetVector<PixelDigi>::const_iterator DSViter = digis->begin(); DSViter != digis->end();
        DSViter++) {
     unsigned int rawid = DSViter->id;
     edm::LogInfo("Phase2TrackerMonitorDigi") << " Det Id = " << rawid;
+
     int layer = tTopo->getITPixelLayerNumber(rawid);
+
+    //std::cout << " Phase2TrackerMonitorDigi: RawId " << rawid << " Layer " << layer << std::endl;
```
[the remark this answers is not in the document]
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:109]: Done

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc (hunk 4 above)
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:50]: Yes, It is a kind of consecutive digis. Should I change the name of that flag?

Comment (mmusich) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:50]: I would be in favor of that, perhaps `StandAloneClusterizerFlag` or something like that? @sroychow suggestions are welcome.

File: Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py (hunk 14 above)
Comment (gsaha009) [reply Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py:28]: Just wanted to check the missedDigiEta, Pt distributions. For tracks with pT ~ 10 GeV, there were no missed Digis. So, I made that pT cutoff lower to see whether there is any Digi not matched with any simulated tracks.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc (hunk 4 above)
Comment (sroychow) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:50]: Your suggestion of using StandAloneClusterizerFlag is fine from my side. @gsaha009 could you change the name please?

File: Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc (hunk 12 above)
Comment (sroychow) [reply Validation/SiTrackerPhase2V/plugins/Phase2TrackerValidateDigi.cc:1113]: @mmusich This is in my todo list. It is also used in the other modules. I will move this out into a separate file after this PR is merged.

File: Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py (hunk 14 above)
Comment (mmusich) [reply Validation/SiTrackerPhase2V/python/Phase2TrackerValidateDigi_cfi.py:28]: @emiglior do you have any recommendation about the pT cutoff for the matching?

**Comment by jfernan2** [comment]
In this first test and, upon completing all the code comments, I noticed that 2D plots DigiRPosVsZPos, DigiXPosVsYPos, OccupancyInRZ, OccupancyInXY  seem to have moved from TrackerPhase2ITDigi/DigiMonitor/   to   TrackerPhase2OTDigi/DigiMonitor/ and now have some renderization problems (except DigiXPosVsYPos), most likely due to a very large number of bins in both x and y axis.
Could you please review those plots?

File: DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py (hunk 6 above)
Comment (gsaha009) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:72]: > In this first test and, upon completing all the code comments, I noticed that 2D plots DigiRPosVsZPos, DigiXPosVsYPos, OccupancyInRZ, OccupancyInXY seem to have moved from TrackerPhase2ITDigi/DigiMonitor/ to TrackerPhase2OTDigi/DigiMonitor/ and now have some renderization problems (except DigiXPosVsYPos), most likely due to a very large number of bins in both x and y axis.
> Could you please review those plots?

Yes. I think somehow I switched the 2D plots on or off in between to test something. Now all these global histograms are present for both IT & OT Digi-Monitor, but they have very fine binning. So, do you suggest to make less binning? Or, I was thinking to switch off the chargeXYmap 2D histogram. The binning is higher for that layer level histogram too.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc (hunk 4 above)
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:50]: yes, I have changed it.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc (hunk 5 above)
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:745]: Sorry, I missed that. Now I have used `tTopo->pxfBlade(det_id)` for IT. I think it is fixed now.

File: DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc [hunk 17]
```diff
@@ -491,10 +477,11 @@ void Phase2TrackerMonitorDigi::bookLayerHistos(DQMStore::IBooker& ibooker,
     else
       local_mes.NumberOfDigisPerDet = nullptr;
 
-    if (pixelFlag_ || (layer < 4 || layer > 6)) {
+    //For endCap: P-type sensors are present only upto ring 10 for discs 1&2 and upto ring 7 for discs 3,4&5
+    if (pixelFlag_ || (layer < 4 || (layer > 6 && (forDisc12UptoRing10 || forDisc345UptoRing7)))) {
```
[the remark this answers is not in the document]
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/Phase2TrackerMonitorDigi.cc:484]: yes, it can. Just to make sure, is `tTopo->tidDisc(det_id)` replaced by `tTopo->tidWheel(det_id)`? @mmusich

**Comment by mmusich** [comment]
@cmsbuild, please test

File: DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py [hunk 18]
```diff
@@ -61,15 +61,15 @@
...
                             xmin   = cms.double(-0.5),
-                            xmax   = cms.double(49.5),
+                            xmax   = cms.double(15.5),
                             switch = cms.bool(True)
                         ), 
                         TotalNumberOfDigisPerLayerH = cms.PSet(
-                            Nbins = cms.int32(1000),
-                            xmin   = cms.double(-0.5),
-                            xmax   = cms.double(4999.5),
+                            Nbins = cms.int32(10000),
```
Comment (jfernan2) [inline DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: Render cannot support such a large binning I am afraid

Comment (jfernan2) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: Well, it actually does in this case, but there are other 2D plots which are not being displayed due to the large binning used

Comment (mmusich) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: @jfernan2 , being ignorant on the matter, can you shine some light on what's the maximum number of 2D bins that can be rendered correctly?

Comment (jfernan2) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: It is not just a matter of binning but a memory limit: objects beyond 8-10MB produce trouble, specially in Online DQM with the ProtoBuf system which goes LS by LS to display the evolving plot.

In this particular case, you have:
DigiXY: TH2 with 1250x1250 bins -> 6MB
DigiRZ: TH2 with 3000x1250 bins -> 15MB
(ocupancies are worse since they are TProfiles...)
OccupXY: TH2Profile with 1250x1250 bins -> 35MB
OccupRZ: TH2Profile with 3000x1250 bins -> 90MB

so, GUI is only able to render DigiXY at this point...[3]

The JSROOT RAW version seems to be able to render 15MB things... [1] but not always and not stable [2]....
[1] https://cmsweb.cern.ch/dqm/dev/jsroot/index.htm?file=https://cmsweb.cern.ch/dqm/dev/data/browse/ROOT/RelVal/CMSSW_11_2_x/DQM_V0001_R000000001__RelVal_wf23234_0_pr__CMSSW_11_2_X-PRcmssw_31563-9567__DQMIO.root
[2] https://cmsweb.cern.ch/dqm/dev/jsroot/index.htm?file=https://cmsweb.cern.ch/dqm/dev/data/browse/ROOT/RelVal/CMSSW_11_2_x/DQM_V0001_R000000001__RelVal_wf23434_999_pr__CMSSW_11_2_X-PRcmssw_31563-9567__DQMIO.root
[3] https://tinyurl.com/y5w6mkns

Comment (mmusich) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: thx, though to be clear this comment:

> specially in Online DQM with the ProtoBuf system which goes LS by LS to display the evolving plot.

certainly does not apply here.
@gsaha009 I would suggest you to reduce all the 2D maps to 1000x1000 max. The charge map is not really interesting and I would keep that optional (for the standalone wf)

Comment (gsaha009) [reply DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py:70]: Yes, I have reduced the binning of all the 2D histograms. Also, I am keeping 5000 bins (max range 25000) of `TotalNumberOfDigisPerLayer` histogram.

**Comment by mmusich** [comment]
please test

**Comment by jfernan2** [comment]
I am sorry @gsaha009  but OccupancyInXY is still high to not be renderized, 9.3MB... Could you reduce a bit more in order to display it properly?

**Comment by gsaha009** [comment]
> I am sorry @gsaha009 but OccupancyInXY is still high to not be renderized, 9.3MB... Could you reduce a bit more in order to display it properly?

Sure. Could you please tell me this OccupancyInXY for IT or OT? Should I reduce the binning for both?

**Comment by sroychow** [comment]
> > I am sorry @gsaha009 but OccupancyInXY is still high to not be renderized, 9.3MB... Could you reduce a bit more in order to display it properly?
> 
> Sure. Could you please tell me this OccupancyInXY for IT or OT? Should I reduce the binning for both?

The occupancy map in OT has to be reduced( e.g.- https://tinyurl.com/y2qnwbvb)

**Comment by jfernan2** [comment]
In IT OccupancyInXY  seems fine (2.8MB); indeed in IT the leading ME is OccupancyInRZ (6.8MB) which now is fine, in baseline it was not renderized but your change made it slimmer.
You can see the changes of your PR here, please check if resolutions are enough:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-09-25-1100+ae5bf0/39016/dqm-histo-comparison-summary.html

**Comment by gsaha009** [comment]
> In IT OccupancyInXY seems fine (2.8MB); indeed in IT the leading ME is OccupancyInRZ (6.8MB) which now is fine, in baseline it was not renderized but your change made it slimmer.
> You can see the changes of your PR here, please check if resolutions are enough:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-09-25-1100+ae5bf0/39016/dqm-histo-comparison-summary.html

Checked 2 Phase2 wfs and the plots look okay to me. I have reduced the binning of OT_OccupancyInXY and IT_OccupancyInRZ also.

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1
2Ds correctly renderized

**Comment by qliphy** [comment]
+1
